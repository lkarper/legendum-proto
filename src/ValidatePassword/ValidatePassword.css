/* 
    This element is in the DOM for users of assistive technology;
    it must appear in the DOM but not be visible on screen
*/
.ValidatePassword__error-container {
    clip: rect(1px, 1px, 1px, 1px);
    clip-path: inset(50%);
    height: 1px;
    width: 1px;
    margin: -1px;
    overflow: hidden;
    padding: 0;
    position: absolute;
}
