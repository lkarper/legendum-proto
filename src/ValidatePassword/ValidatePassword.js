import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import PasswordChecklist from '../PasswordChecklist/PasswordChecklist';
import './ValidatePassword.css';

const ValidatePassword = (props) => {
    const { 
        passwordErrorMessage, 
        passwordError,
        password,
        setPasswordError,
        setPasswordErrorMessage, 
    } = props;
    
    useEffect(() => {
        const newPasswordError = {};
        const passwordErrorArray = [];

        if (password.length < 8) {
            newPasswordError.tooShort = true;
            passwordErrorArray.push('Password must be at least 8 characters in length.');
        } else {
            newPasswordError.tooShort = false;
        }

        if (password.length > 72) {
            newPasswordError.tooLong = true;
            passwordErrorArray.push('Password must be 72 characters or fewer in length.');
        } else {
            newPasswordError.tooLong = false;
        }

        if (password.startsWith(' ') || password.endsWith(' ')) {
            newPasswordError.endSpaces = true;
            passwordErrorArray.push('Password must not begin or end with a space.');
        } else {
            newPasswordError.endSpaces = false;
        }

        if (!/[a-z]/.test(password)) {
            newPasswordError.lowerCase = true;
            passwordErrorArray.push('Password must contain at least one lowercase character.');
        } else {
            newPasswordError.lowerCase = false;
        }

        if (!/[A-Z]/.test(password)) {
            newPasswordError.upperCase = true;
            passwordErrorArray.push('Password must contain at least one uppercase character.');
        } else {
            newPasswordError.upperCase = false;
        }

        if (!/\d/.test(password)) {
            newPasswordError.number = true;
            passwordErrorArray.push('Password must contain at least one number.');
        } else {
            newPasswordError.number = false;
        }

        if (!/\W/.test(password)) {
            newPasswordError.specialChar = true;
            passwordErrorArray.push('Password must contain at least one special character.');
        } else {
            newPasswordError.specialChar = false;
        }
        setPasswordError(newPasswordError);
        setPasswordErrorMessage(passwordErrorArray);
    }, [password, setPasswordError, setPasswordErrorMessage]);

    return (
        <>
            <PasswordChecklist passwordError={passwordError} />
            <div
                className='ValidatePassword__error-container'
                role="alert"
            >
                <p id="password-error-message">
                    {passwordErrorMessage.length 
                        ? passwordErrorMessage.join(' ') 
                        : 'Password meets all requirements.'
                    }
                </p>
            </div>
        </>
    );
}

ValidatePassword.defaultProps = {
    password: '',
    passwordErrorMessage: [], 
    passwordError: {},
    setPasswordError: () => {},
    setPasswordErrorMessage: () => {},
};

ValidatePassword.propTypes = {
    password: PropTypes.string.isRequired,
    passwordErrorMessage: PropTypes.arrayOf(PropTypes.string).isRequired,
    passwordError: PropTypes.shape({
        tooShort: PropTypes.bool,
        tooLong: PropTypes.bool,
        endSpaces: PropTypes.bool,
        upperCase: PropTypes.bool,
        lowerCase: PropTypes.bool,
        number: PropTypes.bool,
        specialChar: PropTypes.bool,
    }),
    setPasswordError: PropTypes.func.isRequired,
    setPasswordErrorMessage: PropTypes.func.isRequired,
};

export default ValidatePassword;
