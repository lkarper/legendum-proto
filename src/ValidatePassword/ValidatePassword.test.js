import React from 'react';
import ReactDOM from 'react-dom';
import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import ValidatePassword from './ValidatePassword';

describe('ValidatePassword component', () => {
    it('renders the UI without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(<ValidatePassword />, div);
        ReactDOM.unmountComponentAtNode(div);
    });

    it('renders the UI as expected', () => {
        const wrapper = mount(<ValidatePassword />);
        expect(toJSON(wrapper)).toMatchSnapshot();
        wrapper.unmount();
    });
});
