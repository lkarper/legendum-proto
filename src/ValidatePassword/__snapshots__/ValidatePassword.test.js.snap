// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ValidatePassword component renders the UI as expected 1`] = `
<ValidatePassword
  password=""
  passwordError={Object {}}
  passwordErrorMessage={Array []}
  setPasswordError={[Function]}
  setPasswordErrorMessage={[Function]}
>
  <PasswordChecklist
    passwordError={Object {}}
  >
    <div
      className="PasswordChecklist__container"
    >
      <p>
        Password must:
      </p>
      <ul>
        <li>
          Be at least 8 characters in length 
          <span
            className="PasswordChecklist__check"
          >
            ✔
          </span>
        </li>
        <li>
          Be no more than 72 characters in length 
          <span
            className="PasswordChecklist__check"
          >
            ✔
          </span>
        </li>
        <li>
          Not begin or end with a space 
          <span
            className="PasswordChecklist__check"
          >
            ✔
          </span>
        </li>
        <li>
          Contain at least one lowercase letter 
          <span
            className="PasswordChecklist__check"
          >
            ✔
          </span>
        </li>
        <li>
          Contain at least one uppercase letter 
          <span
            className="PasswordChecklist__check"
          >
            ✔
          </span>
        </li>
        <li>
          Contain at least one number 
          <span
            className="PasswordChecklist__check"
          >
            ✔
          </span>
        </li>
        <li>
          Contain at least one special character (e.g. !, @, #, $, etc.) 
          <span
            className="PasswordChecklist__check"
          >
            ✔
          </span>
        </li>
      </ul>
    </div>
  </PasswordChecklist>
  <div
    className="ValidatePassword__error-container"
    role="alert"
  >
    <p
      id="password-error-message"
    >
      Password meets all requirements.
    </p>
  </div>
</ValidatePassword>
`;
